File: Python/160_Intersection_of_Two_Linked_Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        dic = {}
        while headA is not None:
            dic[headA] = 1
            headA = headA.next
        while headB is not None:
            if headB in dic:
                return headB
            headB = headB.next
        return headB

# Another solution : find length first

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        currA = headA
        currB = headB
        lenA, lenB = 0, 0
        while currA:
            lenA += 1
            currA = currA.next
        while currB:
            lenB += 1
            currB = currB.next
        currA = headA
        currB = headB
        if lenA > lenB:
            for i in range(lenA - lenB):
                currA = currA.next
        elif lenB > lenA:
            for i in range(lenB - lenA):
                currB = currB.next
        while currA != currB:
            currA = currA.next
            currB = currB.next
        return currA

# Last Solution (not really understand)


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if (headA==None) or (headB==None):
            return None
        x=headA
        y=headB
        while (x!=y):
            if x==None:
                x=headB
            else:
                x=x.next
            if y==None:
                y=headA
            else:
                y=y.next
        return x

File: Java/160__Intersection_of_Two_Linked_Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        ListNode ptr1 = headA, ptr2 = headB;
        while (ptr1 != null && ptr2 != null) {
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }
        if (ptr1 == null) {
            ptr1 = headB;
            while (ptr2 != null) {
                ptr1 = ptr1.next;
                ptr2 = ptr2.next;
            }
            ptr2 = headA;
            while (ptr1 != null && !ptr1.equals(ptr2)) {
                ptr1 = ptr1.next;
                ptr2 = ptr2.next;
            }
            return ptr1;
        } else {
            ptr2 = headA;
            while (ptr1 != null) {
                ptr1 = ptr1.next;
                ptr2 = ptr2.next;
            }
            ptr1 = headB;
            while (ptr1 != null && !ptr1.equals(ptr2)) {
                ptr1 = ptr1.next;
                ptr2 = ptr2.next;
            }
            return ptr1;
        }

    }
}
